core/libp2p/multi/multiaddress.hpp

```diff
@@ -6,11 +6,136 @@
+     */
+    static FactoryResult createMultiaddress(std::string_view address);
+
+    /**
+     * Construct a multiaddress instance from the bytes
+     * @param bytes to be in that multiaddress
+     */
+    static FactoryResult createMultiaddress(std::shared_ptr<ByteBuffer> bytes);
+
+    /**
+     * Encapsulate a multiaddress to this one, such that:
+     * '/ip4/192.168.0.1' after encapsulation with '/udp/138' becomes
+     * '/ip4/192.168.0.1/udp/138'
+     * @param address - another address to be encapsulated into this one
+     */
+    void encapsulate(const Multiaddress &address);
+
+    /**
+     * Encapsulate a multiaddress from this one, such that:
+     * '/ip4/192.168.0.1/udp/138' after decapsulation with '/udp/' becomes
+     * '/ip4/192.168.0.1'
+     * @param address - another address to be decapsulated from this one
+     * @return true, if such address was found and removed, false otherwise
+     */
+    bool decapsulate(const Multiaddress &address);
+
+    /**
+     * Get the textual representation of the address inside
+     * @return stringified address
+     */
+    std::string_view getStringAddress() const;
+
+    /**
+     * Get the byte representation of the address inside
+     * @return bytes address
+     */
+    const std::vector<uint8_t> &getBytesAddress() const;
+
+    /**
+     * Get peer id of this Multiaddress
+     * @return peer id if exists
+     */
+    boost::optional<std::string> getPeerId() const;
+
+    /**
+     * List of protocols, supported by this Multiaddress
+     */
+    enum class Protocol {
+      kIp4,
+      kIp6,
+      kIpfs,
+      kTcp,
+      kUdp,
+      kDccp,
+      kSctp,
+      kUdt,
+      kUtp,
+      kHttp,
+      kHttps,
+      kWs,
+      kOnion,
+      kWebrtc
+    };
+    /**
+     * Get all values, which are under that protocol in this multiaddress
+     * @param proto to be searched for
+     * @return vector of values, if there is at least one under this protocol,
+     * none otherwise
+     */
+    boost::optional<std::vector<std::string>> getValuesForProtocol(
+        Protocol proto) const;
+
+    bool operator==(const Multiaddress &other) const;
+
+   private:
+    /**
+     * Construct a multiaddress instance from both address and bytes
+     * @param address to be in the multiaddress
+     * @param bytes to be in the multiaddress
+     * @throws invalid_argument, if could not retrieve address' family
+     */
+    Multiaddress(std::string &&address, std::shared_ptr<ByteBuffer> bytes);
+
+    /**
+     * Recalculate peer_id inside this address (the first one, if exists)
+     */
+    void calculatePeerId();
+
+    /**
+     * Convert Protocol enum into a string
+     * @param proto to be converted
+     * @return related string
+     */
+    std::string_view protocolToString(Protocol proto) const;
+
+    std::shared_ptr<ByteBuffer> bytes_;
+    std::string stringified_address_;
+
+    boost::optional<std::string> peer_id_;
```

Comment:
std::string?

Reply:
It is essential that this field is optional, as peer_id may or may not exist, and it's much easier to keep it like that and return